lib/ex_activecampaign/api_v1.ex

```diff
@@ -0,0 +1,236 @@
+
+      iex> ExActivecampaign.ApiV1.clean_url("/profile")
+      "http://localhost:8081/v1/profile"
+
+      iex> ExActivecampaign.ApiV1.clean_url("http://localhost")
+      "http://localhost"
+
+      iex> ExActivecampaign.ApiV1.clean_url("http://localhost:8081/b")
+      "http://localhost:8081/b"
+
+      iex> ExActivecampaign.ApiV1.clean_url("http://localhost:8081")
+      "http://localhost:8081/"
+  """
+  def clean_url(url \\ nil) do
+    url
+    |> endpoint_url
+    |> slash_cleanup
+  end
+
+  defp endpoint_url(endpoint) do
+    case endpoint do
+      nil -> ExActivecampaign.base_url_v1()
+      "" -> ExActivecampaign.base_url_v1()
+      "/" <> _ -> ExActivecampaign.base_url_v1() <> endpoint
+      _ -> endpoint
+    end
+  end
+
+  defp slash_cleanup(url) do
+    url
+    |> String.split(":")
+    |> List.last()
+    |> Integer.parse()
+    |> case do
+      {_, ""} -> url <> "/"
+      _ -> url
+    end
+  end
+
+  @doc """
+  Allows headers to be provided as a list or map, which makes it easier to ensure defaults are set
+
+  ## Examples
+
+      iex> ExActivecampaign.ApiV1.clean_headers(%{})
+      [{"Content-Type", "application/x-www-form-urlencoded"}]
+
+      iex> ExActivecampaign.ApiV1.clean_headers(%{"Content-Type" => "application/xml"})
+      [{"Content-Type", "application/x-www-form-urlencoded"}]
+
+      iex> ExActivecampaign.ApiV1.clean_headers(%{"Authorization" => "Bearer abc123"})
+      [{"Authorization","Bearer abc123"}, {"Content-Type", "application/x-www-form-urlencoded"}]
+
+      iex> ExActivecampaign.ApiV1.clean_headers(%{"Authorization" => "Bearer abc123", "Content-Type" => "application/xml"})
+      [{"Authorization","Bearer abc123"}, {"Content-Type", "application/x-www-form-urlencoded"}]
+
+      iex> ExActivecampaign.ApiV1.clean_headers([])
+      [{"Content-Type", "application/x-www-form-urlencoded"}]
+
+      iex> ExActivecampaign.ApiV1.clean_headers([{"apples", "delicious"}])
+      [{"Content-Type", "application/x-www-form-urlencoded"}, {"apples", "delicious"}]
+
+      iex> ExActivecampaign.ApiV1.clean_headers([{"apples", "delicious"}, {"Content-Type", "application/xml"}])
+      [{"Content-Type", "application/x-www-form-urlencoded"}, {"apples", "delicious"}]
+  """
+  def clean_headers(h) when is_map(h) do
+    h
+    |> Map.put("Content-Type", "application/x-www-form-urlencoded")
+    |> Enum.map(& &1)
+  end
+
+  def clean_headers(h) when is_list(h) do
+    headers =
+      h
+      |> Enum.reject(fn {k, _} -> k == "Content-Type" end)
+
+    case headers do
+      [] -> [{"Content-Type", "application/x-www-form-urlencoded"}]
+      _ -> [{"Content-Type", "application/x-www-form-urlencoded"}] ++ headers
+    end
+  end
+
+  def clean_params(query_params) when query_params == %{}, do: []
+  def clean_params(query_params), do: [{:params, query_params}]
+
+  def handle_response({status, body} = _resp) do
+    case status do
+      200 -> body
+      _ -> %{error_message: "Bad Request"}
```

Comment:
Is it worth logging this or preserving the status code?

Reply:
Aaah yes, this - it's been a while since I've touched this now, but IIRC correctly v1 of the API just blanket responds to any request with a 200 status code, and uses a field in the response to indicate whether the request succeeded or failed (yuk!) I therefore put this here just as a bit of a catchall for truly unexpected situations!

https://www.activecampaign.com/api/example.php?call=contact_sync

Reply:
So yes, logging would be good here as it'll catch any 500-series errors at the ActiveCampaign side which will eliminate a part of enquiries should anything go wrong.

Reply:
```suggestion
    case status do
      200 -> body
      x when x in 400..499 -> %{error_message: "Bad Request"}
      x when x in 500..599 -> %{error_message: "Internal Server Error"}
      _ -> %{error_message: "Unknown Error"}
    end
```

I reckon something like this should be ok a precursor to a refactor which would allow the actual error and status to be interpreted, handled, and logged by the integrating codebase.

Reply:
Yeah that looks good!

Reply:
Thanks! 😄